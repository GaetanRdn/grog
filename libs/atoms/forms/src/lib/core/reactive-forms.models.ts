export type OnChangeFn<ValueType> = (value: ValueType | null) => void;
export type OnTouchedFn = () => void;
