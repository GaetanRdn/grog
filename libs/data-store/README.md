# data-store

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test data-store` to execute the unit tests.
