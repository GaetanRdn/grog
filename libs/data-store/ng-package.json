{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../dist/libs/data-store",
  "lib": {
    "entryFile": "src/index.ts"
  }
}
