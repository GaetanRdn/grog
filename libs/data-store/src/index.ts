export * from './lib/data-store.module';
