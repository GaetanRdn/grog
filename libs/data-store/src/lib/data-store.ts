import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable } from 'rxjs';
import { StoredEntity } from './stored-entity';
import {
  Concrete,
  MandatoryPropertiesOf,
  StorableEntity,
  StoreConfig,
  StoredEntityState,
} from './types';
import { deepCopy } from './utils';

@Injectable({
  providedIn: 'root',
})
export class DataStore<
  EntityType,
  Identifier extends keyof Concrete<EntityType> = keyof MandatoryPropertiesOf<EntityType>
> {
  private _store: BehaviorSubject<StoredEntity<EntityType>[]> =
    new BehaviorSubject<StoredEntity<EntityType>[]>([]);

  private _identifierKeys: (keyof EntityType)[];

  get changed$(): Observable<StorableEntity<EntityType>[]> {
    return this._store.asObservable();
  }

  constructor(private _config: StoreConfig<EntityType>) {
    this._identifierKeys = Array.isArray(this._config.extractIdentifier)
      ? this._config.extractIdentifier
      : [this._config.extractIdentifier];
  }

  public store(...entities: EntityType[]): DataStore<EntityType, Identifier> {
    this._store.next([
      ...this._store.getValue(),
      ...entities.map(
        (entity: EntityType) => new StoredEntity(deepCopy(entity), this)
      ),
    ]);

    return this;
  }

  public get(
    identifier: Pick<EntityType, Identifier>
  ): StorableEntity<EntityType> | null {
    return (
      this._store
        .getValue()
        .find((storedEntity: StoredEntity<EntityType>) =>
          this.sameEntity(identifier, storedEntity.entity)
        ) ?? null
    );
  }

  public delete(identifier: Pick<EntityType, Identifier>): void {
    this.updateState(identifier, 'deleted');
  }

  public update(
    entity: Pick<EntityType, Identifier>,
    change: Partial<EntityType> & MandatoryPropertiesOf<EntityType>
  ): void {
    this.updateState(entity, 'updated', change);
  }

  public partialUpdate(
    entity: Pick<EntityType, Identifier>,
    change: Partial<EntityType>
  ): void {
    this.updateState(entity, 'partial', change);
  }

  public create(entity: MandatoryPropertiesOf<EntityType>): void {
    this._store.next([
      ...this._store.getValue(),
      new StoredEntity(deepCopy(entity) as EntityType, this, 'created'),
    ]);
  }

  private updateState(
    identifier: Pick<EntityType, Identifier>,
    state: StoredEntityState,
    change?: Partial<EntityType>
  ): void {
    this._store.next(
      this._store.getValue().map((storedEntity: StoredEntity<EntityType>) => {
        if (this.sameEntity(identifier, storedEntity.entity)) {
          storedEntity['_state'] = state;

          if (change) {
            storedEntity.change = change;
          }
        }
        return storedEntity;
      })
    );
  }

  private sameEntity(
    identifier: Pick<EntityType, Identifier>,
    entity: EntityType
  ): boolean {
    return this._identifierKeys.every(
      (key: keyof EntityType) => (identifier as EntityType)[key] === entity[key]
    );
  }
}
