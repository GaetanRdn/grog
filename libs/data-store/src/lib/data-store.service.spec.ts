import { TestBed } from '@angular/core/testing';
import { of } from 'rxjs';
import { CRUD_SERVICE_TOKEN } from './crud.service';
import { PersonWithId, PersonWithIdCrudService } from './data-store.mocks';
import { DataStore } from './data-store.service';

describe('DataStoreService', () => {
  afterEach(() => jest.clearAllMocks());

  describe('Simple Entity identified by an id', () => {
    let store: DataStore<PersonWithId>;
    let crud: PersonWithIdCrudService;

    beforeEach(() => {
      TestBed.configureTestingModule({
        providers: [
          DataStore,
          // PersonWithIdCrudService,
          {
            provide: CRUD_SERVICE_TOKEN,
            useClass: PersonWithIdCrudService,
          },
        ],
      });

      store = TestBed.inject(DataStore);
      crud = TestBed.inject(CRUD_SERVICE_TOKEN) as PersonWithIdCrudService;
    });

    test('When call get once then check crud has been called once', async () => {
      // GIVEN
      jest
        .spyOn(crud, 'get')
        .mockReturnValueOnce(
          of({ id: 1, firstName: 'Gaetan', lastName: 'Redin' })
        );

      // WHEN
      const personWithId: PersonWithId = await store.get({ id: 1 }).toPromise();

      // THEN
      expect(personWithId).toMatchInlineSnapshot(`
        Object {
          "firstName": "Gaetan",
          "id": 1,
          "lastName": "Redin",
        }
      `);
      expect(crud.get).toHaveBeenCalledTimes(1);
    });

    test('When call get twice then check crud has been called once', async () => {
      // GIVEN
      jest
        .spyOn(crud, 'get')
        .mockReturnValue(of({ id: 1, firstName: 'Gaetan', lastName: 'Redin' }));

      // WHEN
      await store.get({ id: 1 }).toPromise();
      await store.get({ id: 1 }).toPromise();

      // THEN
      expect(crud.get).toHaveBeenCalledTimes(1);
    });
  });
});
