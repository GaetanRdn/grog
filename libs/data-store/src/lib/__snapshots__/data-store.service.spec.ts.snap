// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataStoreService when create entity then check changed$ 1`] = `
Array [
  StoredEntity {
    "change": null,
    "entity": Object {
      "firstName": "Gaetan",
      "lastName": "REDIN",
    },
  },
]
`;

exports[`DataStoreService when delete entity then check changed$ 1`] = `
Array [
  StoredEntity {
    "change": null,
    "entity": Object {
      "firstName": "Gaetan",
      "id": 1,
      "lastName": "REDIN",
    },
  },
]
`;

exports[`DataStoreService when get entity and store contains entity then return it 1`] = `
Object {
  "change": null,
  "entity": Object {
    "firstName": "Gaetan",
    "id": 1,
    "lastName": "REDIN",
  },
}
`;

exports[`DataStoreService when partialUpdate entity then check changed$ 1`] = `
Array [
  StoredEntity {
    "change": Object {
      "age": 31,
    },
    "entity": Object {
      "firstName": "Gaetan",
      "id": 1,
      "lastName": "REDIN",
    },
  },
]
`;

exports[`DataStoreService when store 1 entity then check changed$ 1`] = `
Array [
  StoredEntity {
    "change": null,
    "entity": Object {
      "firstName": "Gaetan",
      "lastName": "REDIN",
    },
  },
]
`;

exports[`DataStoreService when store 2 entities at the same time then check changed$ 1`] = `
Array [
  StoredEntity {
    "change": null,
    "entity": Object {
      "age": 25,
      "firstName": "Elisa",
      "lastName": "REDIN",
    },
  },
  StoredEntity {
    "change": null,
    "entity": Object {
      "firstName": "Gaetan",
      "lastName": "REDIN",
    },
  },
]
`;

exports[`DataStoreService when store 2 entities one after another then check last changed$ 1`] = `
Array [
  StoredEntity {
    "change": null,
    "entity": Object {
      "age": 25,
      "firstName": "Elisa",
      "lastName": "REDIN",
    },
  },
  StoredEntity {
    "change": null,
    "entity": Object {
      "firstName": "Gaetan",
      "lastName": "REDIN",
    },
  },
]
`;

exports[`DataStoreService when update entity then check changed$ 1`] = `
Array [
  StoredEntity {
    "change": Object {
      "age": 31,
      "firstName": "Gaetan",
      "id": 1,
      "lastName": "REDIN",
    },
    "entity": Object {
      "firstName": "Gaetan",
      "id": 1,
      "lastName": "REDIN",
    },
  },
]
`;
